File: internal/engine/c_api/api_data/memory_info.cc
/**
 * Copyright 2019 The Gamma Authors.
 *
 * This source code is licensed under the Apache License, Version 2.0 license
 * found in the LICENSE file in the root directory of this source tree.
 */

#include "memory_info.h"

namespace vearch {

MemoryInfo::MemoryInfo() { memory_info_ = nullptr; }

int MemoryInfo::Serialize(char **out, int *out_len) {
  flatbuffers::FlatBufferBuilder builder;
  auto table = gamma_api::CreateMemoryInfo(
      builder, table_mem_bytes_, index_mem_bytes_, vector_mem_bytes_,
      field_range_mem_bytes_, bitmap_mem_bytes_);
  builder.Finish(table);
  *out_len = builder.GetSize();
  *out = (char *)malloc(*out_len * sizeof(char));
  memcpy(*out, (char *)builder.GetBufferPointer(), *out_len);
  return 0;
}

void MemoryInfo::Deserialize(const char *data, int len) {
  memory_info_ =
      const_cast<gamma_api::MemoryInfo *>(gamma_api::GetMemoryInfo(data));

  table_mem_bytes_ = memory_info_->table_mem();
  index_mem_bytes_ = memory_info_->index_mem();
  vector_mem_bytes_ = memory_info_->vector_mem();
  field_range_mem_bytes_ = memory_info_->field_range_mem();
  bitmap_mem_bytes_ = memory_info_->bitmap_mem();
}

long MemoryInfo::TableMem() { return table_mem_bytes_; }

void MemoryInfo::SetTableMem(long table_mem) { table_mem_bytes_ = table_mem; }

long MemoryInfo::IndexMem() { return index_mem_bytes_; }

void MemoryInfo::SetIndexMem(long index_mem_bytes) {
  index_mem_bytes_ = index_mem_bytes;
}

long MemoryInfo::VectorMem() { return vector_mem_bytes_; }

void MemoryInfo::SetVectorMem(long vector_mem_bytes) {
  vector_mem_bytes_ = vector_mem_bytes;
}

long MemoryInfo::FieldRangeMem() { return field_range_mem_bytes_; }

void MemoryInfo::SetFieldRangeMem(long field_range_mem_bytes) {
  field_range_mem_bytes_ = field_range_mem_bytes;
}

long MemoryInfo::BitmapMem() { return bitmap_mem_bytes_; }

void MemoryInfo::SetBitmapMem(long bitmap_mem_bytes) {
  bitmap_mem_bytes_ = bitmap_mem_bytes;
}

}  // namespace vearch

File: internal/engine/sdk/go/gamma/memory_info.go
/**
 * Copyright 2019 The Vearch Authors.
 *
 * This source code is licensed under the Apache License, Version 2.0 license
 * found in the LICENSE file in the root directory of this source tree.
 */

package gamma

import (
	flatbuffers "github.com/google/flatbuffers/go"
	"github.com/vearch/vearch/v3/internal/engine/idl/fbs-gen/go/gamma_api"
)

type MemoryInfo struct {
	TableMem      int64
	IndexMem      int64
	VectorMem     int64
	FieldRangeMem int64
	BitmapMem     int64

	memoryInfo *gamma_api.MemoryInfo
}

func (status *MemoryInfo) Serialize() []byte {
	builder := flatbuffers.NewBuilder(0)

	gamma_api.MemoryInfoStart(builder)
	gamma_api.MemoryInfoAddTableMem(builder, status.TableMem)
	gamma_api.MemoryInfoAddIndexMem(builder, status.IndexMem)
	gamma_api.MemoryInfoAddVectorMem(builder, status.VectorMem)
	gamma_api.MemoryInfoAddFieldRangeMem(builder, status.FieldRangeMem)
	gamma_api.MemoryInfoAddBitmapMem(builder, status.BitmapMem)
	builder.Finish(builder.EndObject())
	return builder.FinishedBytes()
}

func (status *MemoryInfo) DeSerialize(buffer []byte) {
	status.memoryInfo = gamma_api.GetRootAsMemoryInfo(buffer, 0)
	status.TableMem = status.memoryInfo.TableMem()
	status.IndexMem = status.memoryInfo.IndexMem()
	status.VectorMem = status.memoryInfo.VectorMem()
	status.FieldRangeMem = status.memoryInfo.FieldRangeMem()
	status.BitmapMem = status.memoryInfo.BitmapMem()
}
